Implement BFBT preconditioner

quangx opened the pull request:
This adds the weighted BFBT preconditioner to ASPECT (currently work in progress).

tjhei commented on benchmarks/nsinker/current.prm:
```diff
@@ -0,0 +1,11 @@
+subsection Material model
```
remove this file and .clangd again.

tjhei commented on benchmarks/nsinker/nsinker.prm:
```diff
@@ -13,14 +13,16 @@ set Nonlinear solver scheme                = no Advection, iterated Stokes
     set Maximum number of expensive Stokes solver steps = 1000
     set Linear solver tolerance                         = 1e-6
     set GMRES solver restart length                     = 100
+    set Use weighted BFBT for Schur complement	        = true
```
turn off "locally conervative" a few lines down.

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -142,9 +143,57 @@ namespace aspect
             }
+          else if (this->get_parameters().use_full_A_block_preconditioner==false && this->get_parameters().use_bfbt) //not full a and bfbt
+            {
+              double sqrt_eta=sqrt(eta);
```
const

tjhei replied:
make this const

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -142,9 +143,57 @@ namespace aspect
+                    {
+                      data.local_matrix(i, j) += ((2.0 * eta * (scratch.grads_phi_u[i]
+                                                                * scratch.grads_phi_u[j]))
+                                                  + one_over_eta * pressure_scaling
```
1/sqrt

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -142,9 +143,57 @@ namespace aspect
+
+                        data.local_matrix(i, j) += (one_over_eta * pressure_scaling
+                                                    * pressure_scaling
+                                                    * (scratch.phi_p[i]
```
grad_p

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -124,8 +126,8 @@ namespace aspect
           const double JxW = scratch.finite_element_values.JxW(q);
-
-          if (this->get_parameters().use_full_A_block_preconditioner == false)
+//assembles stokes preconditioner ? qxh
```
remove comment

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -124,8 +126,8 @@ namespace aspect
-          if (this->get_parameters().use_full_A_block_preconditioner == false)
+//assembles stokes preconditioner ? qxh
+          if (this->get_parameters().use_full_A_block_preconditioner == false && !this->get_parameters().use_bfbt) //not full a and not bfbt
             {
```
We need to split up the assembly code into something sensible. I think we could split the computation of the velocity and the pressure block (at least within this assembler).

tjhei replied:
@gassmoeller ?

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -142,9 +144,62 @@ namespace aspect
+                {
+                  for (unsigned int j = 0; j < stokes_dofs_per_cell; ++j)
+                    {
+                      // if (scratch.dof_component_indices[i] ==
```
put back in

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -142,9 +144,62 @@ namespace aspect
+            }
+          else if (this->get_parameters().use_full_A_block_preconditioner && this->get_parameters().use_bfbt) //full a and bfbt
+            {
+              double sqrt_eta=sqrt(eta);
```
const

tjhei commented on source/simulator/assembly.cc:
```diff
@@ -333,6 +334,7 @@ namespace aspect
     current_constraints.distribute_local_to_global (data.local_matrix,
                                                     data.local_dof_indices,
                                                     system_preconditioner_matrix);
+    current_constraints.distribute_local_to_global(data.local_lumped_mass_matrix,data.local_dof_indices,lumped_mass_matrix);
```
only do if bfbt

tjhei commented on source/simulator/assembly.cc:
```diff
@@ -396,6 +398,26 @@ namespace aspect
          StokesPreconditioner<dim> (stokes_dofs_per_cell));
 
     system_preconditioner_matrix.compress(VectorOperation::add);
+    lumped_mass_matrix.compress(VectorOperation::add);
```
only do if bfbt

tjhei commented on source/simulator/assembly.cc:
```diff
@@ -396,6 +398,26 @@ namespace aspect
+        else
+          {
+
+            if (lumped_mass_matrix.block(0)[i] == 0.0)
```
remove

tjhei commented on source/simulator/core.cc:
```diff
@@ -1276,6 +1276,7 @@ namespace aspect
       }
 
     system_preconditioner_matrix.reinit (sp);
+    lumped_mass_matrix.reinit(introspection.index_sets.stokes_partitioning);
```
only do if bfbt

tjhei commented on source/simulator/solver.cc:
```diff
@@ -368,7 +368,129 @@ namespace aspect
+    {
+      // Trilinos reports a breakdown in case src=dst=0, even though it should return
+      // convergence without iterating. We simply skip solving in this case.
+        if(src.l2_norm() > 1e-50){
```
not necessary. Remove.

tjhei commented on source/simulator/solver.cc:
```diff
@@ -368,7 +368,129 @@ namespace aspect
+            {
+              TrilinosWrappers::MPI::Vector utmp;
+              utmp.reinit(lumped_mass_matrix);
+              utmp=0;
```
these are not necessary, as reinit() zeros things out

tjhei commented on source/simulator/solver.cc:
```diff
@@ -368,7 +368,129 @@ namespace aspect
+
+                utmp.scale(lumped_mass_matrix); // D^{-1}   B^T(BD^{-1}B^T}^{-1})y2  stored in utmp
+
+                system_matrix.block(0,0).vmult(wtmp,utmp); //A   D^{-1}B^T(BD^{-1}B^T}^{-1})y2 stored in wtmp
+                wtmp.scale(lumped_mass_matrix); //C^{-1}  AD^{-1}B^T(BD^{-1}B^T}^{-1})y2 stored in wtmp
+
+                system_matrix.block(1,0).vmult(ptmp,wtmp); //B  C^{-1}AD^{-1}B^T(BD^{-1}B^T}^{-1})y2 stored in ptmp
```
```suggestion
                system_matrix.block(0,0).vmult(wtmp,utmp); // A * ...

                wtmp.scale(lumped_mass_matrix);  // C *...

                system_matrix.block(1,0).vmult(ptmp,wtmp);   //B *...
```

tjhei replied:
this is still open...

tjhei commented on source/simulator/solver.cc:
```diff
@@ -368,7 +368,129 @@ namespace aspect
+    unsigned int WeightedBFBT<PreconditionerMp>::n_iterations() const
+    {
+      return n_iterations_;
+    }
```
three blank lines

tjhei commented on source/simulator/solver.cc:
```diff
@@ -927,12 +1049,10 @@ namespace aspect
 
         solver_control_cheap.enable_history_data();
         solver_control_expensive.enable_history_data();
-
```
add these back

tjhei commented on source/simulator/solver.cc:
```diff
@@ -977,11 +1125,20 @@ namespace aspect
-                          distributed_stokes_solution,
-                          distributed_stokes_rhs,
-                          preconditioner_cheap);
+            if (!parameters.use_bfbt)
```
remove the ! and swap blocks

tjhei commented on source/simulator/solver.cc:
```diff
@@ -1025,11 +1182,22 @@ namespace aspect
                 solver(solver_control_expensive, mem,
                        SolverFGMRES<LinearAlgebra::BlockVector>::
                        AdditionalData(number_of_temporary_vectors));
+                if (!parameters.use_bfbt)
```
same here

tjhei commented on source/simulator/solver.cc:
```diff
@@ -1042,11 +1210,22 @@ namespace aspect
-                                           inverse_velocity_block_cheap.n_iterations()+inverse_velocity_block_expensive.n_iterations(),
-                                           solver_control_cheap,
-                                           solver_control_expensive);
+                if (!parameters.use_bfbt)
```
same here

tjhei commented:
please add a test (copy tests/nsinker.prm)

bangerth commented:
Please let me know when you've addressed @tjhei 's comments and I can take a look too.

tjhei commented on benchmarks/nsinker/nsinker.prm:
```diff
@@ -13,19 +13,25 @@ set Nonlinear solver scheme                = no Advection, single Stokes
 # Follow as closely as possible the parameters from Rudi et al. (2017)
 subsection Solver parameters
-  subsection Stokes solver parameters
+   subsection Stokes solver parameters
```
can you undo this change?

tjhei commented on benchmarks/nsinker/nsinker.prm:
```diff
@@ -13,19 +13,25 @@ set Nonlinear solver scheme                = no Advection, single Stokes
 set Max nonlinear iterations               = 1
 set Use years in output instead of seconds = false
 
+
```
can you add a new nsinker-bfbt.prm instead of changing this file?

tjhei commented on benchmarks/nsinker/nsinker.prm:
```diff
@@ -13,19 +13,25 @@ set Nonlinear solver scheme                = no Advection, single Stokes
   end
+  subsection AMG parameters
+   set AMG aggregation threshold = 0.02
+	end
```
indentation

tjhei commented on include/aspect/simulator.h:
```diff
@@ -2088,6 +2088,7 @@ namespace aspect
        * solving.
        */
       LinearAlgebra::BlockSparseMatrix                          system_matrix;
+      LinearAlgebra::BlockVector                                lumped_mass_matrix;
```
Can you add a comment before this line that this vector is used for BFBT?

tjhei commented on source/simulator/assemblers/interface.cc:
```diff
@@ -53,8 +53,9 @@ namespace aspect
+          phi_u(stokes_dofs_per_cell,numbers::signaling_nan<Tensor<1,dim>>()),
           phi_p_c (add_compaction_pressure ? stokes_dofs_per_cell : 0, numbers::signaling_nan<double>()),
-          grad_phi_p (add_compaction_pressure ? stokes_dofs_per_cell : 0, numbers::signaling_nan<Tensor<1,dim>>()),
+          grad_phi_p (stokes_dofs_per_cell, numbers::signaling_nan<Tensor<1,dim>>()),
```
```suggestion
          grad_phi_p ((add_compaction_pressure || use_bfbt) ? stokes_dofs_per_cell : 0, numbers::signaling_nan<Tensor<1,dim>>()),
```
but you need to pass use_bfbt to the constructor.

tjhei commented on source/simulator/assemblers/interface.cc:
```diff
@@ -78,6 +79,7 @@ namespace aspect
           grads_phi_u (scratch.grads_phi_u),
           div_phi_u (scratch.div_phi_u),
           phi_p (scratch.phi_p),
+          phi_u(scratch.phi_u),
```
```suggestion
          phi_u (scratch.phi_u),
```

gassmoeller replied:
```suggestion
          phi_u (scratch.phi_u),
```

tjhei commented on source/simulator/assemblers/interface.cc:
```diff
@@ -389,6 +392,7 @@ namespace aspect
         StokesPreconditioner (const StokesPreconditioner &data)
           :
           local_matrix (data.local_matrix),
+          local_lumped_mass_matrix(data.local_lumped_mass_matrix),
```
```suggestion
          local_lumped_mass_matrix (data.local_lumped_mass_matrix),
```

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -125,38 +130,152 @@ namespace aspect
 
-          if (this->get_parameters().use_full_A_block_preconditioner == false)
-            {
+          if(this->get_parameters().use_full_A_block_preconditioner==false){
```
```suggestion
          if(this->get_parameters().use_full_A_block_preconditioner==false)
          {
```

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -125,38 +130,152 @@ namespace aspect
+                    if (scratch.dof_component_indices[i] ==
+                        scratch.dof_component_indices[j])
+                      {
+                        data.local_matrix(i, j) += ((2.0 * eta * (scratch.grads_phi_u[i]
+                                                                  * scratch.grads_phi_u[j]))
+                                                    )
```
remove the outer-most ()

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -125,38 +130,152 @@ namespace aspect
+                                                    )
+                                                  * JxW;
+                      }
+            
```
remove empty lines?

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -125,38 +130,152 @@ namespace aspect
+                      {
+
+
+                            // i and j are not pressures
```
remove comment and empty lines

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -125,38 +130,152 @@ namespace aspect
+                            data.local_lumped_mass_matrix[i]+=sqrt_eta*scalar_product(scratch.phi_u[i],scratch.phi_u[j])*JxW;
+
+
+      // i and j are pressures?
```
remove comment and empty lines

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -125,38 +130,152 @@ namespace aspect
+
+      // i and j are pressures?
+                          if(scratch.dof_component_indices[i]==pressure_component_index && scratch.dof_component_indices[j]==pressure_component_index)
+                          data.local_matrix(i, j) += (
```
you did not run indent

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -125,38 +130,152 @@ namespace aspect
+      // i and j are pressures?
+                          if(scratch.dof_component_indices[i]==pressure_component_index && scratch.dof_component_indices[j]==pressure_component_index)
+                          data.local_matrix(i, j) += (
+                                                      1.0/sqrt_eta * pressure_scaling
+                                                      * pressure_scaling
```
```suggestion
                                                      1.0/sqrt_eta * pressure_scaling * pressure_scaling
```

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -125,38 +130,152 @@ namespace aspect
+                                                    * JxW;
+                       }
+                                                      
+                          
```
get rid of excessive empty lines, please

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -125,38 +130,152 @@ namespace aspect
+                
+
+          }
+          else{
```
```suggestion
          else
          {
```
braces always go on separate lines in ASPECT.

tjhei commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -125,38 +130,152 @@ namespace aspect
+          //                                         + one_over_eta * pressure_scaling
+          //                                         * pressure_scaling
+          //                                         * (scratch.phi_p[i]
+          //                                            * scratch.phi_p[j]))
```
remove all these comments

quangx commented:
> Please let me know when you've addressed @tjhei 's comments and I can take a look too.

I addressed the the above comments and am working on the newer ones

gassmoeller commented on source/simulator/assemblers/interface.cc:
```diff
@@ -379,6 +381,7 @@ namespace aspect
           :
           local_matrix (stokes_dofs_per_cell,
                         stokes_dofs_per_cell),
+          local_inverse_lumped_mass_matrix(stokes_dofs_per_cell),
```
```suggestion
          local_inverse_lumped_mass_matrix (stokes_dofs_per_cell),
```

gassmoeller commented on source/simulator/assemblers/interface.cc:
```diff
@@ -389,6 +392,7 @@ namespace aspect
         StokesPreconditioner (const StokesPreconditioner &data)
           :
           local_matrix (data.local_matrix),
+          local_inverse_lumped_mass_matrix(data.local_inverse_lumped_mass_matrix),
```
```suggestion
          local_inverse_lumped_mass_matrix (data.local_inverse_lumped_mass_matrix),
```

gassmoeller commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -125,7 +130,7 @@ namespace aspect
           const double JxW = scratch.finite_element_values.JxW(q);
 
-          if (this->get_parameters().use_full_A_block_preconditioner == false)
+          if (this->get_parameters().use_full_A_block_preconditioner==false)
```
```suggestion
          if (this->get_parameters().use_full_A_block_preconditioner == false)
```

gassmoeller commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -134,29 +139,146 @@ namespace aspect
+                        data.local_inverse_lumped_mass_matrix[i]+=sqrt_eta*scalar_product(scratch.phi_u[i],scratch.phi_u[j])*JxW;
+
+
+                      // i and j are pressures?
```
be consistent about the question mark here and in line 159. Either both need it, or none.

gassmoeller commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -134,29 +139,146 @@ namespace aspect
-          else
+          if (this->get_parameters().use_bfbt)
             {
+              const double sqrt_eta=sqrt(eta);
```
I see you prefer to not have spaces around assignments or comparisons. However, most of our code uses spaces. We sometimes allow both forms for loop counters (`unsigned int i=0; i<x; ++i`) but usually we have space around assignment and comparisons to make the code more readable. Please check your PR for other instances.

gassmoeller commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -134,29 +139,146 @@ namespace aspect
                                                    * JxW;
-                      }
+                    }
+
+
+
+
```
Are these empty lines necessary?

gassmoeller commented on source/simulator/assemblers/stokes.cc:
```diff
@@ -134,29 +139,146 @@ namespace aspect
+
+
+                }
+
+
```
same question as above

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -396,6 +399,26 @@ namespace aspect
+              }
+            else
+              {
+
+
```
remove empty lines

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -252,6 +253,8 @@ namespace aspect
       * This operator can either just apply the preconditioner (AMG)
       * or perform an inner CG solve with the same preconditioner.
       */
+
+
```
remove empty lines

gassmoeller replied:
empty lines

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -367,8 +370,135 @@ namespace aspect
+            solver.solve(mp_matrix,
+                         ptmp,
+                         src,
+                         mp_preconditioner); //(BD^{-1}B^T}^{-1})y2 store in ptmp
```
We do not use inline comments. Please move all comments in this function before the line they are describing.

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -367,8 +370,135 @@ namespace aspect
+            wtmp.scale(inverse_lumped_mass_matrix); //C^{-1}  AD^{-1}B^T(BD^{-1}B^T}^{-1})y2 stored in wtmp
+
+            system_matrix.block(1,0).vmult(ptmp,wtmp); //B  C^{-1}AD^{-1}B^T(BD^{-1}B^T}^{-1})y2 stored in ptmp
+
+
+
```
maybe one or two empty lines is sufficient?

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -955,12 +1102,24 @@ namespace aspect
+                                        *schur,
                                         inverse_velocity_block_expensive);
 
+
+
+
+
+
+        // create an expensive preconditioner that solves for the A block with CG
+
+
+
+
+
+
```
I think the number of empty lines here is a bit excessive ;-). Reduce to one or at most two.

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -983,6 +1142,9 @@ namespace aspect
                           distributed_stokes_rhs,
                           preconditioner_cheap);
 
+
+
+
```
same here adding this many empty lines does not follow our usual code style.

As a more general comment: If you find code that would want to improve with changes that are unrelated to your PR topic, please do not add the changes to the current PR, but instead open an new one.

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -1026,10 +1188,15 @@ namespace aspect
+                              distributed_stokes_solution,
+                              distributed_stokes_rhs,
+                              preconditioner_expensive);
+
+
+
+
+
```
remove empty lines

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -1042,12 +1209,17 @@ namespace aspect
             // processors
             catch (const std::exception &exc)
               {
+
+
```
remove empty lines

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -1042,12 +1209,17 @@ namespace aspect
                                            solver_control_cheap,
                                            solver_control_expensive);
 
+
+
+
```
remove empty lines

gassmoeller commented on tests/nsinker_bfbt.prm:
```diff
@@ -0,0 +1,14 @@
+set Dimension = 3
+
+include $ASPECT_SOURCE_DIR/benchmarks/nsinker/nsinker.prm
+
+subsection Solver parameters
+   subsection Stokes solver parameters
+   set Use weighted BFBT for Schur complement          = true
+
+  end
+end
+subsection Mesh refinement
+  set Initial adaptive refinement              = 0
+  set Initial global refinement                = 1
+end
```
Like Timo commented above: keep this test, but also make a new nsinker_bfbt benchmark .prm that looks more or less like this test (including the original benchmark and switching to bfbt).

gassmoeller reviewed (commented):
Very nice improvement! I think the structure of the code is fine, I only have a bunch of style comments. And also you need to add a changelog entry, see [here](https://github.com/geodynamics/aspect/blob/main/CONTRIBUTING.md#changelog-entries) for how to add one.

tjhei commented on doc/modules/changes/20240606_hoang:
```diff
@@ -0,0 +1,5 @@
+New: We implement an option to use the Weighted BFBT preconditioner 
+introduced by Rudi et al (2017). This preconditioner was designed 
+for problems involving highly heteregenous viscosities.
```
heterogeneous

tjhei commented on source/simulator/assemblers/interface.cc:
```diff
@@ -116,15 +119,18 @@ namespace aspect
                                      add_compaction_pressure,
-                                     rebuild_stokes_matrix),
+                                     rebuild_stokes_matrix,
+                                     use_bfbt
```
put brace on same line

tjhei commented on source/simulator/assembly.cc:
```diff
@@ -794,7 +818,9 @@ namespace aspect
-                            assemble_newton_stokes_matrix),
+                            assemble_newton_stokes_matrix,
+                            use_bfbt
+                            ),
```
put ) on same line

quangx commented:
ready to be reviewed again

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -391,11 +394,30 @@ namespace aspect
+    if (parameters.use_bfbt)
+      {
+        inverse_lumped_mass_matrix.compress(VectorOperation::add);
+        IndexSet local_indices=inverse_lumped_mass_matrix.block(0).locally_owned_elements();
```
I asked you in a previous review to add spaces around assignment operators ` = `. Here and below they are still missing. You can go through your own pull request and search to these cases here: https://github.com/geodynamics/aspect/pull/5753/files

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -29,6 +29,7 @@
 #include <deal.II/lac/solver_gmres.h>
 #include <deal.II/lac/solver_bicgstab.h>
 #include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/linear_operator.h>
```
Timo's comment to remove the linear operator include is not resolved yet

gassmoeller reviewed (commented):
I am not going to finish this review yet. The code is almost ready, but please go through this page from top to bottom and check for every comment if you have resolved and pushed them yet: https://github.com/geodynamics/aspect/pull/5753/files

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -798,7 +918,6 @@ namespace aspect
         // for a direct solver, we have only one block for velocity+pressure,
         // and so only one block needs to be copied.
         solution.block(block_vel_and_pres) = distributed_stokes_solution.block(0);
-
```
keep this empty line

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -1043,17 +1175,16 @@ namespace aspect
                                            inverse_velocity_block_cheap.n_iterations()+inverse_velocity_block_expensive.n_iterations(),
                                            solver_control_cheap,
                                            solver_control_expensive);
-
```
Keep this empty line

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -1043,17 +1175,16 @@ namespace aspect
+
                 if (parameters.n_expensive_stokes_solver_steps > 0)
                   solver_controls.push_back(solver_control_expensive);
-
```
Keep this empty line

gassmoeller reviewed (commented):
Some smaller commits.

tjhei commented on source/simulator/assembly.cc:
```diff
@@ -391,11 +394,30 @@ namespace aspect
+          {
+            if (current_constraints.is_constrained(i))
+              {
+                inverse_lumped_mass_matrix.block(0)[i] = 1;
```
```suggestion
                inverse_lumped_mass_matrix.block(0)[i] = 1.0;
```

tjhei commented on source/simulator/solver.cc:
```diff
@@ -367,6 +366,135 @@ namespace aspect
+         * @param mp_preconditioner The preconditioner for @p mp_matrix
+         * @param solver_tolerance The relative solver tolerance for the inner solve
+         * @param inverse_lumped_mass_matrix Lumped mass matrix associated with the velocity block
+         */
```
you missed the last argument

tjhei commented on source/simulator/solver.cc:
```diff
@@ -367,6 +366,135 @@ namespace aspect
+    void WeightedBFBT<PreconditionerMp>::vmult(TrilinosWrappers::MPI::Vector &dst,
+                                               const TrilinosWrappers::MPI::Vector &src) const
+    {
+      // Trilinos reports a breakdown in case src=dst=0, even though it should return
+      // convergence without iterating. We simply skip solving in this case.
```
remove this comment

tjhei commented on source/simulator/solver.cc:
```diff
@@ -928,10 +1056,24 @@ namespace aspect
+          }
+        else
+          {
+            schur=std::make_unique<internal::InverseWeightedMassMatrix<TrilinosWrappers::PreconditionBase>>(
```
```suggestion
            schur = std::make_unique<internal::InverseWeightedMassMatrix<TrilinosWrappers::PreconditionBase>>(
```

tjhei commented on source/simulator/solver.cc:
```diff
@@ -955,12 +1097,12 @@ namespace aspect
+                                        *schur,
                                         inverse_velocity_block_expensive);
-
+        // create an expensive preconditioner that solves for the A block with CG
```
remove this line

tjhei reviewed (approved):
I think this is close to done. Great job!

tjhei reviewed (approved):
Looks good to go from my side.

gassmoeller commented:
Congrats :tada:
